### PR TITLE
Feature/webhooks

### PR DESCRIPTION
**Pull request by msadovyi**

### CODE REVIEW

**Review by valcker — APPROVED** [review]

File: packages/gateway/src/server/router.ts [hunk 1]
```diff
@@ -165,6 +167,8 @@ export async function updateApi(apiDef: ApiDef): Promise<void> {
   const routerWithNewApi = Router();
   await buildApi(routerWithNewApi, apiDef, mesh);
 
+  webhooks.triggerForApi(apiDef.name, WebhookEvents.SCHEMA_CHANGED);
```
Comment (germankrause) [inline packages/gateway/src/server/router.ts:170]: trigger after replacing api in router

File: packages/gateway/src/webhooks/Webhooks.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import { promisify } from 'util';
+import Events from 'events';
+import { prefixLogger } from '@graphql-portal/logger';
+import { ApiDef, WebhookEvents } from '@graphql-portal/types';
+
+export type RemoveEventListener = (() => void) | null;
+
+class Webhooks {
+  private readonly hooks: {
+    [key: string]: {
```
Comment (germankrause) [inline packages/gateway/src/webhooks/Webhooks.ts:11]: apiName: string

File: packages/gateway/src/webhooks/Webhooks.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      // key equal to api name
+      [key: string]: {
```
Comment (germankrause) [inline packages/gateway/src/webhooks/Webhooks.ts:13]: eventName: string

File: packages/gateway/src/webhooks/Webhooks.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        // key equal to event name
+        [key: string]: RemoveEventListener; // key equal to url
```
Comment (germankrause) [inline packages/gateway/src/webhooks/Webhooks.ts:15]: url: string

File: packages/types/src/enum/webhook-events.enum.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export enum WebhookEvents {
```
Comment (germankrause) [inline packages/types/src/enum/webhook-events.enum.ts:1]: WebhookEvent

Comment (germankrause) [reply packages/types/src/enum/webhook-events.enum.ts:1]: as enum represents one of the value variants

File: packages/types/src/enum/index.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
 export * from './channel.enum';
 export * from './metrics.enum';
 export * from './tracers.enum';
+export * from './webhook-events.enum';
```
Comment (germankrause) [inline packages/types/src/enum/index.ts:4]: webhook-event.enum

File: packages/types/src/api-def-schema.json [hunk 7]
```diff
@@ -1530,6 +1531,23 @@
...
+        "additionalProperties": false,
+        "type": "object",
+        "title": "Webhook",
+        "properties": {
+          "event": {
+            "type": "string",
+            "enum": ["schema_changed"],
+            "description": "Name of the event you want to subscribe"
+          },
+          "url": { "type": "string", "description": "URL that remote API will ping" }
```
Comment (germankrause) [inline packages/types/src/api-def-schema.json:1546]: add regex validation for url

**Review by germankrause — CHANGES_REQUESTED** [review]
update dependencies

File: packages/gateway/src/webhooks/Webhooks.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import https from 'https';
+import { promisify } from 'util';
+import Events from 'events';
+import { prefixLogger } from '@graphql-portal/logger';
+import { ApiDef, WebhookEvents } from '@graphql-portal/types';
+
+export type RemoveEventListener = (() => void) | null;
+
+class Webhooks {
```
Comment (germankrause) [inline packages/gateway/src/webhooks/Webhooks.ts:9]: add unit tests

**Pull request merged**